README.md
# socket.io-client-dart

Port of awesome JavaScript Node.js library - [Socket.io-client v2.0.1](https://github.com/socketio/socket.io-client) - in Dart

## Usage

**Dart Server**
```dart
import 'package:socket_io/socket_io.dart';
import 'package:socket_io_client_flutter/socket_io_client.dart' as IO;

main() {
  // Dart server
  var io = new Server();
  var nsp = io.of('/some');
  nsp.on('connection', (client) {
    print('connection /some');
    client.on('msg', (data) {
      print('data from /some => $data');
      client.emit('fromServer', "ok 2");
    });
  });
  io.on('connection', (client) {
    print('connection default namespace');
    client.on('msg', (data) {
      print('data from default => $data');
      client.emit('fromServer', "ok");
    });
  });
  io.listen(3000);
}
```
**Dart Client**
```dart

import 'package:socket_io_client/socket_io_client.dart' as IO;

main() {
  // Dart client
  IO.Socket socket = IO.io('http://localhost:3000');
  socket.onConnect((_) {
    print('connect');
    socket.emit('msg', 'test');
  });
  socket.on('event', (data) => print(data));
  socket.onDisconnect((_) => print('disconnect'));
  socket.on('fromServer', (_) => print(_));
}
```

### Connect manually

To connect the socket manually, set the option `autoConnect: false` and call `.connect()`.

For example,

<pre>
Socket socket = io('http://localhost:3000', 
    OptionBuilder()
      .setTransports(['websocket']) // for Flutter or Dart VM
      .<b>disableAutoConnect()</b>  // disable auto-connection
      .setExtraHeaders({'foo': 'bar'}) // optional
      .build()
  );
<b>socket.connect();</b>
</pre>

Note that `.connect()` should not be called if `autoConnect: true` 
(by default, it's enabled to true), as this will cause all event handlers to get registered/fired twice. See [Issue #33](https://github.com/rikulo/socket.io-client-dart/issues/33).

### Update the extra headers

```dart
Socket socket = ... // Create socket.
socket.io.options['extraHeaders'] = {'foo': 'bar'}; // Update the extra headers.
socket.io..disconnect()..connect(); // Reconnect the socket manually.
```

### Emit with acknowledgement

```dart
Socket socket = ... // Create socket.
socket.onConnect((_) {
    print('connect');
    socket.emitWithAck('msg', 'init', ack: (data) {
        print('ack $data') ;
        if (data != null) {
          print('from server $data');
        } else {
          print("Null") ;
        }
    });
});
```

### Socket connection events

These events can be listened on.

```dart
const List EVENTS = [
  'connect',
  'connect_error',
  'connect_timeout',
  'connecting',
  'disconnect',
  'error',
  'reconnect',
  'reconnect_attempt',
  'reconnect_failed',
  'reconnect_error',
  'reconnecting',
  'ping',
  'pong'
];

// Replace 'onConnect' with any of the above events.
socket.onConnect((_) {
    print('connect');
});
```

### Acknowledge with the socket server that an event has been received.

```dart
socket.on('eventName', (data) {
    final dataList = data as List;
    final ack = dataList.last as Function;
    ack(null);
});
```

## Usage (Flutter)

In Flutter env. not (Flutter Web env.) it only works with `dart:io` websocket,
 not with `dart:html` websocket or Ajax (XHR), so in this case
you have to add `setTransports(['websocket'])` when creates the socket instance.

For example,

```dart
IO.Socket socket = IO.io('http://localhost:3000',
  OptionBuilder()
      .setTransports(['websocket']) // for Flutter or Dart VM
      .setExtraHeaders({'foo': 'bar'}) // optional
      .build());
```

## Usage with stream and streambuilder in Flutter

```dart
import 'dart:async';


// STEP1:  Stream setup
class StreamSocket{
  final _socketResponse= StreamController<String>();

  void Function(String) get addResponse => _socketResponse.sink.add;

  Stream<String> get getResponse => _socketResponse.stream;

  void dispose(){
    _socketResponse.close();
  }
}

StreamSocket streamSocket =StreamSocket();

//STEP2: Add this function in main function in main.dart file and add incoming data to the stream
void connectAndListen(){
  IO.Socket socket = IO.io('http://localhost:3000',
      OptionBuilder()
       .setTransports(['websocket']).build());

    socket.onConnect((_) {
     print('connect');
     socket.emit('msg', 'test');
    });

    //When an event recieved from server, data is added to the stream
    socket.on('event', (data) => streamSocket.addResponse);
    socket.onDisconnect((_) => print('disconnect'));

}

//Step3: Build widgets with streambuilder

class BuildWithSocketStream extends StatelessWidget {
  const BuildWithSocketStream({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Container(
      child: StreamBuilder(
        stream: streamSocket.getResponse ,
        builder: (BuildContext context, AsyncSnapshot<String> snapshot){
          return Container(
            child: snapshot.data,
          );
        },
      ),
    );
  }
}

```

## Troubleshooting

### Cannot connect "https" server or self-signed certificate server

- Refer to https://github.com/dart-lang/sdk/issues/34284 issue.
  The workround is to use the following code provided by [@lehno](https://github.com/lehno) on [#84](https://github.com/rikulo/socket.io-client-dart/issues/84)

```dart
class MyHttpOverrides extends HttpOverrides {
  @override
  HttpClient createHttpClient(SecurityContext context) {
    return super.createHttpClient(context)
      ..badCertificateCallback =
          (X509Certificate cert, String host, int port) => true;
  }
}

void main() {
  HttpOverrides.global = new MyHttpOverrides();
  runApp(MyApp());
}
```

### Memory leak issues in iOS when closing socket.

- Refer to https://github.com/rikulo/socket.io-client-dart/issues/108 issue.
  Please use `socket.dispose()` instead of `socket.close()` or `socket.disconnect()` to solve the memory leak issue on iOS.

### Connect_error on MacOS with SocketException: Connection failed
* Refer to https://github.com/flutter/flutter/issues/47606#issuecomment-568522318 issue.
           
By adding the following key into the to file `*.entitlements` under directory `macos/Runner/`
```
<key>com.apple.security.network.client</key>
<true/>
```

For more details, please take a look at https://flutter.dev/desktop#setting-up-entitlements

## Notes to Contributors

### Fork socket.io-client-dart

If you'd like to contribute back to the core, you can [fork this repository](https://help.github.com/articles/fork-a-repo) and send us a pull request, when it is ready.

If you are new to Git or GitHub, please read [this guide](https://help.github.com/) first.

## Who Uses

- [Quire](https://quire.io) - a simple, collaborative, multi-level task management tool.
- [KEIKAI](https://keikai.io/) - a web spreadsheet for Big Data.

## Socket.io Dart Server

- [socket.io-dart](https://github.com/rikulo/socket.io-dart)

## Contributors

- Thanks [@felangel](https://github.com/felangel) for https://github.com/rikulo/socket.io-client-dart/issues/7
- Thanks [@Oskang09](https://github.com/Oskang09) for https://github.com/rikulo/socket.io-client-dart/issues/21
- Thanks [@bruce3x](https://github.com/bruce3x) for https://github.com/rikulo/socket.io-client-dart/issues/25
- Thanks [@Kavantix](https://github.com/Kavantix) for https://github.com/rikulo/socket.io-client-dart/issues/26
- Thanks [@luandnguyen](https://github.com/luandnguyen) for https://github.com/rikulo/socket.io-client-dart/issues/59


CHANGELOG.md
## 1.0.0-nullsafety.0

**New Feature:**

* [#48](https://github.com/DanielWHoward/socket_io_client_flutter/issues/1) change code for Flutter

## 1.0.0-nullsafety.0

**New Feature:**

* [#132](https://github.com/rikulo/socket.io-client-dart/issues/132) Migrating to null safety for Dart


## 0.9.12

**New Feature:**

* [#46](https://github.com/rikulo/socket.io-client-dart/issues/46) Make this library more "Darty"

## 0.9.11

**New Feature:**

* [#108](https://github.com/rikulo/socket.io-client-dart/issues/108) Need dispose method for clearing resources

## 0.9.10+2

* Fix dart analyzer warning for formatting issues.

## 0.9.10+1

* Fix dart analyzer warning.

## 0.9.10

**Bug fix:**

* [#72](https://github.com/rikulo/socket.io-client-dart/issues/72) Can't send Iterable as one packet

## 0.9.9

**Bug fix:**

* [#67](https://github.com/rikulo/socket.io-client-dart/issues/67) Retry connection backoff after 54 tries reconnections every 0 second

## 0.9.8

**Bug fix:**

* [#33](https://github.com/rikulo/socket.io-client-dart/issues/33) socket.on('receiveMessage',(data)=>print("data")) called twice


## 0.9.7+2

**New Feature:**

* [#48](https://github.com/rikulo/socket.io-client-dart/issues/48) add links to github repo in pubspec.yaml


## 0.9.7+1

**New Feature:**

* [#38](https://github.com/rikulo/socket.io-client-dart/issues/38) Improve pub.dev score


## 0.9.6+3

**Bug fix:**

* [#42](https://github.com/rikulo/socket.io-client-dart/issues/42) Error when using emitWithAck

## 0.9.5

**New Feature:**

* [#34](https://github.com/rikulo/socket.io-client-dart/issues/34) Add support for extraHeaders

**Bug fix:**

* [#39](https://github.com/rikulo/socket.io-client-dart/issues/39) The factor of Backoff with 54 retries causes an overflow


lib/src/socket.dart
import 'dart:typed_data';

///
/// socket.dart
///
/// Purpose:
///
/// Description:
///
/// History:
///   26/04/2017, Created by jumperchen
///
/// Copyright (C) 2017 Potix Corporation. All Rights Reserved.
import 'package:logging/logging.dart';
import 'package:socket_io_common/src/util/event_emitter.dart';
import '../src/manager.dart';
import '../src/on.dart' as util;
import 'package:socket_io_common/src/parser/parser.dart';

///
/// Internal events (blacklisted).
/// These events can't be emitted by the user.
///
/// @api private
///

const List EVENTS = [
  'connect',
  'connect_error',
  'connect_timeout',
  'connecting',
  'disconnect',
  'error',
  'reconnect',
  'reconnect_attempt',
  'reconnect_failed',
  'reconnect_error',
  'reconnecting',
  'ping',
  'pong'
];

final Logger _logger = Logger('socket_io_client:Socket');

///
/// `Socket` constructor.
///
/// @api public
class Socket extends EventEmitter {
  String nsp;
  Map opts;

  Manager io;
  Socket json;
  num ids = 0;
  Map acks = {};
  bool connected = false;
  bool disconnected = true;
  List sendBuffer = [];
  List receiveBuffer = [];
  String query;
  List subs;
  Map flags;
  String id;

  Socket(this.io, this.nsp, this.opts) {
    json = this; // compat
    if (opts != null) {
      query = opts['query'];
    }
    if (io.autoConnect) open();
  }

  ///
  /// Subscribe to open, close and packet events
  ///
  /// @api private
  void subEvents() {
    if (subs?.isEmpty == true) return;
//     if (subs?.isNotEmpty == true) return;

    var io = this.io;
    subs = [
      util.on(io, 'open', onopen),
      util.on(io, 'packet', onpacket),
      util.on(io, 'close', onclose)
    ];
  }

  ///
  /// "Opens" the socket.
  ///
  /// @api public
  Socket open() => connect();

  Socket connect() {
    if (connected == true) return this;
    subEvents();
    io.open(); // ensure open
    if ('open' == io.readyState) onopen();
    emit('connecting');
    return this;
  }

  ///
  /// Sends a `message` event.
  ///
  /// @return {Socket} self
  /// @api public
  Socket send(List args) {
    emit('message', args);
    return this;
  }

  ///
  /// Override `emit`.
  /// If the event is in `events`, it's emitted normally.
  ///
  /// @param {String} event name
  /// @return {Socket} self
  /// @api public
  @override
  void emit(String event, [data]) {
    emitWithAck(event, data);
  }

  void emitWithBinary(String event, [data]) {
    emitWithAck(event, data, binary: true);
  }

  ///
  /// Emits to this client.
  ///
  /// @return {Socket} self
  /// @api public
  void emitWithAck(String event, dynamic data,
      {Function ack, bool binary = false}) {
    if (EVENTS.contains(event)) {
      super.emit(event, data);
    } else {
      var sendData = <dynamic>[event];
      if (data is ByteBuffer || data is List<int>) {
        sendData.add(data);
      } else if (data is Iterable) {
        sendData.addAll(data);
      } else if (data != null) {
        sendData.add(data);
      }

      var packet = {
        'type': binary ? BINARY_EVENT : EVENT,
        'data': sendData,
        'options': {'compress': flags?.isNotEmpty == true && flags['compress']}
      };

      // event ack callback
      if (ack != null) {
        _logger.fine('emitting packet with ack id $ids');
        acks['${ids}'] = ack;
        packet['id'] = '${ids++}';
      }

      if (connected == true) {
        this.packet(packet);
      } else {
        sendBuffer.add(packet);
      }
      flags = null;
    }
  }

  ///
  /// Sends a packet.
  ///
  /// @param {Object} packet
  /// @api private
  void packet(Map packet) {
    packet['nsp'] = nsp;
    io.packet(packet);
  }

  ///
  /// Called upon engine `open`.
  ///
  /// @api private
  void onopen([_]) {
    _logger.fine('transport is open - connecting');

    // write connect packet if necessary
    if ('/' != nsp) {
      if (query?.isNotEmpty == true) {
        packet({'type': CONNECT, 'query': query});
      } else {
        packet({'type': CONNECT});
      }
    }
  }

  ///
  /// Called upon engine `close`.
  ///
  /// @param {String} reason
  /// @api private
  void onclose(reason) {
    _logger.fine('close ($reason)');
    emit('disconnecting', reason);
    connected = false;
    disconnected = true;
    id = null;
    emit('disconnect', reason);
  }

  ///
  /// Called with socket packet.
  ///
  /// @param {Object} packet
  /// @api private
  void onpacket(packet) {
    if (packet['nsp'] != nsp) return;

    switch (packet['type']) {
      case CONNECT:
        onconnect();
        break;

      case EVENT:
        onevent(packet);
        break;

      case BINARY_EVENT:
        onevent(packet);
        break;

      case ACK:
        onack(packet);
        break;

      case BINARY_ACK:
        onack(packet);
        break;

      case DISCONNECT:
        ondisconnect();
        break;

      case ERROR:
        emit('error', packet['data']);
        break;
    }
  }

  ///
  /// Called upon a server event.
  ///
  /// @param {Object} packet
  /// @api private
  void onevent(Map packet) {
    List args = packet['data'] ?? [];
//    debug('emitting event %j', args);

    if (null != packet['id']) {
//      debug('attaching ack callback to event');
      args.add(ack(packet['id']));
    }

    // dart doesn't support "String... rest" syntax.
    if (connected == true) {
      if (args.length > 2) {
        Function.apply(super.emit, [args.first, args.sublist(1)]);
      } else {
        Function.apply(super.emit, args);
      }
    } else {
      receiveBuffer.add(args);
    }
  }

  ///
  /// Produces an ack callback to emit with an event.
  ///
  /// @api private
  Function ack(id) {
    var sent = false;
    return (_) {
      // prevent double callbacks
      if (sent) return;
      sent = true;
      _logger.fine('sending ack $_');

      packet({
        'type': ACK,
        'id': id,
        'data': [_]
      });
    };
  }

  ///
  /// Called upon a server acknowlegement.
  ///
  /// @param {Object} packet
  /// @api private
  void onack(Map packet) {
    var ack = acks.remove(packet['id']);
    if (ack is Function) {
      _logger.fine('''calling ack ${packet['id']} with ${packet['data']}''');

      var args = packet['data'] as List;
      if (args.length > 1) {
        // Fix for #42 with nodejs server
        Function.apply(ack, [args]);
      } else {
        Function.apply(ack, args);
      }
    } else {
      _logger.fine('''bad ack ${packet['id']}''');
    }
  }

  ///
  /// Called upon server connect.
  ///
  /// @api private
  void onconnect() {
    connected = true;
    disconnected = false;
    emit('connect');
    emitBuffered();
  }

  ///
  /// Emit buffered events (received and emitted).
  ///
  /// @api private
  void emitBuffered() {
    var i;
    for (i = 0; i < receiveBuffer.length; i++) {
      List args = receiveBuffer[i];
      if (args.length > 2) {
        Function.apply(super.emit, [args.first, args.sublist(1)]);
      } else {
        Function.apply(super.emit, args);
      }
    }
    receiveBuffer = [];

    for (i = 0; i < sendBuffer.length; i++) {
      packet(sendBuffer[i]);
    }
    sendBuffer = [];
  }

  ///
  /// Called upon server disconnect.
  ///
  /// @api private
  void ondisconnect() {
    _logger.fine('server disconnect (${nsp})');
    destroy();
    onclose('io server disconnect');
  }

  ///
  /// Called upon forced client/server side disconnections,
  /// this method ensures the manager stops tracking us and
  /// that reconnections don't get triggered for this.
  ///
  /// @api private.

  void destroy() {
    if (subs?.isNotEmpty == true) {
      // clean subscriptions to avoid reconnections
      for (var i = 0; i < subs.length; i++) {
        subs[i].destroy();
      }
      subs = null;
    }

    io.destroy(this);
  }

  ///
  /// Disconnects the socket manually.
  ///
  /// @return {Socket} self
  /// @api public
  Socket close() => disconnect();

  Socket disconnect() {
    if (connected == true) {
      _logger.fine('performing disconnect (${nsp})');
      packet({'type': DISCONNECT});
    }

    // remove socket from pool
    destroy();

    if (connected == true) {
      // fire events
      onclose('io client disconnect');
    }
    return this;
  }

  /// Disposes the socket manually which will destroy, close, disconnect the socket connection
  /// and clear all the event listeners. Unlike [close] or [disconnect] which won't clear
  /// all the event listeners
  ///
  /// @since 0.9.11
  void dispose() {
    disconnect();
    clearListeners();
  }

  ///
  /// Sets the compress flag.
  ///
  /// @param {Boolean} if `true`, compresses the sending data
  /// @return {Socket} self
  /// @api public
  Socket compress(compress) {
    flags = flags ??= {};
    flags['compress'] = compress;
    return this;
  }
}


lib/socket_io_client.dart
///
/// socket_io_client.dart
///
/// Purpose:
///
/// Description:
///
/// History:
///   26/04/2017, Created by jumperchen
///
/// Copyright (C) 2017 Potix Corporation. All Rights Reserved.
///

library socket_io_client;

import 'package:logging/logging.dart';
import './src/socket.dart';
import './src/engine/parser/parser.dart' as parser;
import './src/engine/parseqs.dart';
import './src/manager.dart';

export './src/socket.dart';
export './src/darty.dart';

// Protocol version
final protocol = parser.protocol;

final Map<String, dynamic> cache = {};

final Logger _logger = Logger('socket_io_client');

///
/// Looks up an existing `Manager` for multiplexing.
/// If the user summons:
///
///   `io('http://localhost/a');`
///   `io('http://localhost/b');`
///
/// We reuse the existing instance based on same scheme/port/host,
/// and we initialize sockets for each namespace.
///
/// @api public
///
Socket io(uri, [opts]) => _lookup(uri, opts);

Socket _lookup(uri, opts) {
  opts = opts ?? <dynamic, dynamic>{};

  var parsed = Uri.parse(uri);
  var id = '${parsed.scheme}://${parsed.host}:${parsed.port}';
  var path = parsed.path;
  var sameNamespace = cache.containsKey(id) && cache[id].nsps.containsKey(path);
  var newConnection = opts['forceNew'] == true ||
      opts['force new connection'] == true ||
      false == opts['multiplex'] ||
      sameNamespace;

  var io;

  if (newConnection) {
    _logger.fine('ignoring socket cache for $uri');
    io = Manager(uri: uri, options: opts);
  } else {
    io = cache[id] ??= Manager(uri: uri, options: opts);
  }
  if (parsed.query.isNotEmpty && opts['query'] == null) {
    opts['query'] = parsed.query;
  } else if (opts != null && opts['query'] is Map) {
    opts['query'] = encode(opts['query']);
  }
  String nsp = parsed.path.isEmpty ? '/' : parsed.path;
  return io.socket(nsp, opts);
}


lib/src/manager.dart
/// Copyright (C) 2017 Potix Corporation. All Rights Reserved
/// History: 2017-04-26 15:27
/// Author: jumperchen<jumperchen@potix.com>
import 'dart:async';
import 'dart:math' as math;

import 'package:logging/logging.dart';
import 'package:socket_io_common/src/util/event_emitter.dart';
import 'package:socket_io_common/src/parser/parser.dart';
import '../src/on.dart';
import '../src/socket.dart';
import '../src/engine/socket.dart' as engine_socket;
import '../src/on.dart' as util;

final Logger _logger = Logger('socket_io_client:Manager');

///
/// `Manager` constructor.
///
/// @param {String} engine instance or engine uri/opts
/// @param {Object} options
/// @api public
///
class Manager extends EventEmitter {
  // Namespaces
  Map<String, Socket> nsps = {};
  List subs = [];
  Map options;

  bool _reconnection;
  num _reconnectionAttempts;
  num _reconnectionDelay;
  num _randomizationFactor;
  num _reconnectionDelayMax;
  num _timeout;
  _Backoff backoff;
  String readyState = 'closed';
  String uri;
  List connecting = [];
  num lastPing;
  bool encoding = false;
  List packetBuffer = [];
  bool reconnecting = false;

  engine_socket.Socket engine;
  Encoder encoder = Encoder();
  Decoder decoder = Decoder();
  bool autoConnect;
  bool skipReconnect;

  Manager({uri, Map options}) {
    options = options ?? <dynamic, dynamic>{};

    options['path'] ??= '/socket.io';
    this.options = options;
    reconnection = options['reconnection'] != false;
    reconnectionAttempts = options['reconnectionAttempts'] ?? double.infinity;
    reconnectionDelay = options['reconnectionDelay'] ?? 1000;
    reconnectionDelayMax = options['reconnectionDelayMax'] ?? 5000;
    randomizationFactor = options['randomizationFactor'] ?? 0.5;
    backoff = _Backoff(
        min: reconnectionDelay,
        max: reconnectionDelayMax,
        jitter: randomizationFactor);
    timeout = options['timeout'] ?? 20000;
    this.uri = uri;
    autoConnect = options['autoConnect'] != false;
    if (autoConnect) open();
  }

  ///
  /// Propagate given event to sockets and emit on `this`
  ///
  /// @api private
  ///
  void emitAll(String event, [data]) {
    emit(event, data);
    for (var nsp in nsps.keys) {
      nsps[nsp].emit(event, data);
    }
  }

  ///
  /// Update `socket.id` of all sockets
  ///
  /// @api private
  ///
  void updateSocketIds() {
    for (var nsp in nsps.keys) {
      nsps[nsp].id = generateId(nsp);
    }
  }

  ///
  /// generate `socket.id` for the given `nsp`
  ///
  /// @param {String} nsp
  /// @return {String}
  /// @api private
  ///
  String generateId(String nsp) {
    if (nsp.startsWith('/')) nsp = nsp.substring(1);
    return (nsp.isEmpty ? '' : (nsp + '#')) + (engine.id ?? '');
  }

  ///
  /// Sets the `reconnection` config.
  ///
  /// @param {Boolean} true/false if it should automatically reconnect
  /// @return {Manager} self or value
  /// @api public
  ///
  bool get reconnection => _reconnection;
  set reconnection(bool v) => _reconnection = v;

  ///
  /// Sets the reconnection attempts config.
  ///
  /// @param {Number} max reconnection attempts before giving up
  /// @return {Manager} self or value
  /// @api public
  ///
  num get reconnectionAttempts => _reconnectionAttempts;
  set reconnectionAttempts(num v) => _reconnectionAttempts = v;

  ///
  /// Sets the delay between reconnections.
  ///
  /// @param {Number} delay
  /// @return {Manager} self or value
  /// @api public
  ///
  num get reconnectionDelay => _reconnectionDelay;
  set reconnectionDelay(num v) => _reconnectionDelay = v;

  num get randomizationFactor => _randomizationFactor;
  set randomizationFactor(num v) {
    _randomizationFactor = v;
    backoff?.jitter = v;
  }

  ///
  /// Sets the maximum delay between reconnections.
  ///
  /// @param {Number} delay
  /// @return {Manager} self or value
  /// @api public
  ///
  num get reconnectionDelayMax => _reconnectionDelayMax;
  set reconnectionDelayMax(num v) {
    _reconnectionDelayMax = v;
    backoff?.max = v;
  }

  ///
  /// Sets the connection timeout. `false` to disable
  ///
  /// @return {Manager} self or value
  /// @api public
  ///
  num get timeout => _timeout;
  set timeout(num v) => _timeout = v;

  ///
  /// Starts trying to reconnect if reconnection is enabled and we have not
  /// started reconnecting yet
  ///
  /// @api private
  ///
  void maybeReconnectOnOpen() {
    // Only try to reconnect if it's the first time we're connecting
    if (!reconnecting && _reconnection == true && backoff.attempts == 0) {
      // keeps reconnection from firing twice for the same reconnection loop
      reconnect();
    }
  }

  ///
  /// Sets the current transport `socket`.
  ///
  /// @param {Function} optional, callback
  /// @return {Manager} self
  /// @api public
  ///
  Manager open({callback, Map opts}) =>
      connect(callback: callback, opts: opts);

  Manager connect({callback, Map opts}) {
    _logger.fine('readyState $readyState');
    if (readyState.contains('open')) return this;

    _logger.fine('opening $uri');
    engine = engine_socket.Socket(uri, options);
    var socket = engine;
    readyState = 'opening';
    skipReconnect = false;

    // emit `open`
    var openSub = util.on(socket, 'open', (_) {
      onopen();
      if (callback != null) callback();
    });

    // emit `connect_error`
    var errorSub = util.on(socket, 'error', (data) {
      _logger.fine('connect_error');
      cleanup();
      readyState = 'closed';
      emitAll('connect_error', data);
      if (callback != null) {
        callback({'error': 'Connection error', 'data': data});
      } else {
        // Only do this if there is no fn to handle the error
        maybeReconnectOnOpen();
      }
    });

    // emit `connect_timeout`
    if (_timeout != null) {
      var timeout = _timeout;
      _logger.fine('connect attempt will timeout after $timeout');

      // set timer
      var timer = Timer(Duration(milliseconds: timeout.toInt()), () {
        _logger.fine('connect attempt timed out after $timeout');
        openSub.destroy();
        socket.close();
        socket.emit('error', 'timeout');
        emitAll('connect_timeout', timeout);
      });

      subs.add(Destroyable(() => timer?.cancel()));
    }

    subs.add(openSub);
    subs.add(errorSub);

    return this;
  }

  ///
  /// Called upon transport open.
  ///
  /// @api private
  ///
  void onopen([_]) {
    _logger.fine('open');

    // clear old subs
    cleanup();

    // mark as open
    readyState = 'open';
    emit('open');

    // add subs
    var socket = engine;
    subs.add(util.on(socket, 'data', ondata));
    subs.add(util.on(socket, 'ping', onping));
    subs.add(util.on(socket, 'pong', onpong));
    subs.add(util.on(socket, 'error', onerror));
    subs.add(util.on(socket, 'close', onclose));
    subs.add(util.on(decoder, 'decoded', ondecoded));
  }

  ///
  /// Called upon a ping.
  ///
  /// @api private
  ///
  void onping([_]) {
    lastPing = DateTime.now().millisecondsSinceEpoch;
    emitAll('ping');
  }

  ///
  /// Called upon a packet.
  ///
  /// @api private
  ///
  void onpong([_]) {
    emitAll('pong', DateTime.now().millisecondsSinceEpoch - lastPing);
  }

  ///
  /// Called with data.
  ///
  /// @api private
  ///
  void ondata(data) {
    decoder.add(data);
  }

  ///
  /// Called when parser fully decodes a packet.
  ///
  /// @api private
  ///
  void ondecoded(packet) {
    emit('packet', packet);
  }

  ///
  /// Called upon socket error.
  ///
  /// @api private
  ///
  void onerror(err) {
    _logger.fine('error $err');
    emitAll('error', err);
  }

  ///
  /// Creates a socket for the given `nsp`.
  ///
  /// @return {Socket}
  /// @api public
  ///
  Socket socket(String nsp, Map opts) {
    var socket = nsps[nsp];

    var onConnecting = ([_]) {
      if (!connecting.contains(socket)) {
        connecting.add(socket);
      }
    };

    if (socket == null) {
      socket = Socket(this, nsp, opts);
      nsps[nsp] = socket;
      socket.on('connecting', onConnecting);
      socket.on('connect', (_) {
        socket.id = generateId(nsp);
      });

      if (autoConnect) {
        // manually call here since connecting event is fired before listening
        onConnecting();
      }
    }

    return socket;
  }

  ///
  /// Called upon a socket close.
  ///
  /// @param {Socket} socket
  ///
  void destroy(socket) {
    connecting.remove(socket);
    if (connecting.isNotEmpty) return;

    close();
  }

  ///
  /// Writes a packet.
  ///
  /// @param {Object} packet
  /// @api private
  ///
  void packet(Map packet) {
    _logger.fine('writing packet $packet');
    if (packet.containsKey('query') && packet['type'] == 0) {
      packet['nsp'] += '''?${packet['query']}''';
    }

    if (encoding != true) {
      // encode, then write to engine with result
      encoding = true;
      encoder.encode(packet, (encodedPackets) {
        for (var i = 0; i < encodedPackets.length; i++) {
          engine.write(encodedPackets[i], packet['options']);
        }
        encoding = false;
        processPacketQueue();
      });
    } else {
      // add packet to the queue
      packetBuffer.add(packet);
    }
  }

  ///
  /// If packet buffer is non-empty, begins encoding the
  /// next packet in line.
  ///
  /// @api private
  ///
  void processPacketQueue() {
    if (packetBuffer.isNotEmpty && encoding != true) {
      var pack = packetBuffer.removeAt(0);
      packet(pack);
    }
  }

  ///
  /// Clean up transport subscriptions and packet buffer.
  ///
  /// @api private
  ///
  void cleanup() {
    _logger.fine('cleanup');

    var subsLength = subs.length;
    for (var i = 0; i < subsLength; i++) {
      var sub = subs.removeAt(0);
      sub.destroy();
    }

    packetBuffer = [];
    encoding = false;
    lastPing = null;

    decoder.destroy();
  }

  ///
  /// Close the current socket.
  ///
  /// @api private
  ///
  void close() => disconnect();

  void disconnect() {
    _logger.fine('disconnect');
    skipReconnect = true;
    reconnecting = false;
    if ('opening' == readyState) {
      // `onclose` will not fire because
      // an open event never happened
      cleanup();
    }
    backoff.reset();
    readyState = 'closed';
    engine.close();
  }

  ///
  /// Called upon engine close.
  ///
  /// @api private
  ///
  void onclose(error) {
    _logger.fine('onclose');

    cleanup();
    backoff.reset();
    readyState = 'closed';
    emit('close', error['reason']);

    if (_reconnection == true && !skipReconnect) {
      reconnect();
    }
  }

  ///
  /// Attempt a reconnection.
  ///
  /// @api private
  ///
  Manager reconnect() {
    if (reconnecting || skipReconnect) return this;

    if (backoff.attempts >= _reconnectionAttempts) {
      _logger.fine('reconnect failed');
      backoff.reset();
      emitAll('reconnect_failed');
      reconnecting = false;
    } else {
      var delay = backoff.duration;
      _logger.fine('will wait %dms before reconnect attempt', delay);

      reconnecting = true;
      var timer = Timer(Duration(milliseconds: delay.toInt()), () {
        if (skipReconnect) return;

        _logger.fine('attempting reconnect');
        emitAll('reconnect_attempt', backoff.attempts);
        emitAll('reconnecting', backoff.attempts);

        // check again for the case socket closed in above events
        if (skipReconnect) return;

        open(callback: ([err]) {
          if (err != null) {
            _logger.fine('reconnect attempt error');
            reconnecting = false;
            reconnect();
            emitAll('reconnect_error', err['data']);
          } else {
            _logger.fine('reconnect success');
            onreconnect();
          }
        });
      });

      subs.add(Destroyable(() => timer.cancel()));
    }
    return this;
  }

  ///
  /// Called upon successful reconnect.
  ///
  /// @api private
  ///
  void onreconnect() {
    var attempt = backoff.attempts;
    reconnecting = false;
    backoff.reset();
    updateSocketIds();
    emitAll('reconnect', attempt);
  }
}

///
/// Initialize backoff timer with `opts`.
///
/// - `min` initial timeout in milliseconds [100]
/// - `max` max timeout [10000]
/// - `jitter` [0]
/// - `factor` [2]
///
/// @param {Object} opts
/// @api public
class _Backoff {
  num _ms;
  num _max;
  final num _factor;
  num _jitter;
  num attempts = 0;

  _Backoff({min = 100, max = 10000, jitter = 0, factor = 2})
      : _ms = min,
        _max = max,
        _factor = factor {
    _jitter = jitter > 0 && jitter <= 1 ? jitter : 0;
  }

  ///
  /// Return the backoff duration.
  ///
  /// @return {Number}
  /// @api public
  ///
  num get duration {
    var ms = _ms * math.pow(_factor, attempts++);
    if (_jitter > 0) {
      var rand = math.Random().nextDouble();
      var deviation = (rand * _jitter * ms).floor();
      ms = ((rand * 10).floor() & 1) == 0 ? ms - deviation : ms + deviation;
    }
    // #39: avoid an overflow with negative value
    ms = math.min(ms, _max);
    return ms <= 0 ? _max : ms;
  }

  ///
  /// Reset the number of attempts.
  ///
  /// @api public
  ///
  void reset() {
    attempts = 0;
  }

  ///
  /// Set the minimum duration
  ///
  /// @api public
  ///
  set min(min) => _ms = min;

  ///
  /// Set the maximum duration
  ///
  /// @api public
  ///
  set max(max) => _max = max;

  ///
  /// Set the jitter
  ///
  /// @api public
  ///
  set jitter(jitter) => _jitter = jitter;
}
